aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -89,52 +92,136 @@ m_getObjectRetries(0),
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     Aws::Map<Aws::String, Aws::String>&& metadata,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
+                                     bool createBucket,
+                                     bool doConsistencyChecks) :
+UploadFileRequest("", fileStream, bucketName, keyName, contentType, std::move(metadata), s3Client, executor, createBucket, doConsistencyChecks, true)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     const Aws::Map<Aws::String, Aws::String>& metadata,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
                                      bool createBucket,
                                      bool doConsistencyChecks) :
-UploadFileRequest(fileName, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, createBucket, doConsistencyChecks)
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, executor, createBucket, doConsistencyChecks)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
+                                     bool createBucket,
+                                     bool doConsistencyChecks) :
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>{}, s3Client, executor, createBucket, doConsistencyChecks)
 {
 }
 
 UploadFileRequest::~UploadFileRequest()
 {
-    m_fileStream.close();
+    m_executor->WaitForCompletion();
+    if (auto fileStream = std::dynamic_pointer_cast<Aws::FStream>(m_inputStream))

```

Comment:
dynamic cast uses RTTI, which we turn off on some builds. Just do a static cast or a c-style cast.